File: engine/src/main/java/de/hanno/hpengine/engine/transform/Spatial.java
package de.hanno.hpengine.engine.transform;

import de.hanno.hpengine.engine.camera.Camera;
import org.joml.Vector3f;
import org.joml.Vector3fc;

import static java.lang.Float.MAX_VALUE;

public interface Spatial {
    Vector3fc MIN = new Vector3f(MAX_VALUE, MAX_VALUE, MAX_VALUE);
    Vector3fc MAX = new Vector3f(-MAX_VALUE, -MAX_VALUE, -MAX_VALUE);

    static boolean isInFrustum(Camera camera, Vector3f centerWorld, Vector3f minWorld, Vector3f maxWorld) {
        Vector3f tempDistVector = new Vector3f();
        new Vector3f(minWorld).sub(maxWorld, tempDistVector);

//		if (de.hanno.hpengine.camera.getFrustum().pointInFrustum(minWorld.x, minWorld.y, minWorld.z) ||
//			de.hanno.hpengine.camera.getFrustum().pointInFrustum(maxWorld.x, maxWorld.y, maxWorld.z)) {
//		if (de.hanno.hpengine.camera.getFrustum().cubeInFrustum(cubeCenterX, cubeCenterY, cubeCenterZ, size)) {
//		if (de.hanno.hpengine.camera.getFrustum().pointInFrustum(minView.x, minView.y, minView.z)
//				|| de.hanno.hpengine.camera.getFrustum().pointInFrustum(maxView.x, maxView.y, maxView.z)) {
        if (camera.getFrustum().sphereInFrustum(centerWorld.x, centerWorld.y, centerWorld.z, tempDistVector.length()/2)) {
            return true;
        }
        return false;
    }

    Vector3f getCenterWorld(Transform transform);

    AABB getMinMaxWorld(Transform transform);

    AABB getMinMax();

    float getBoundingSphereRadius(Transform transform);

    Vector3f getCenterWorld();

    float getBoundingSphereRadius();

    AABB getMinMaxWorld();

    default void update(float deltaSeconds) {}
}

File: engine/src/main/java/de/hanno/hpengine/engine/transform/Spatial.kt
package de.hanno.hpengine.engine.transform

import de.hanno.hpengine.engine.model.AnimatedModel
import de.hanno.hpengine.engine.model.StaticModel
import org.joml.Vector3f
import java.io.Serializable

open class SimpleSpatial(override val boundingVolume: AABB = AABB(Vector3f(Spatial.MIN),Vector3f(Spatial.MAX))) : Serializable, Spatial

inline val SimpleSpatial.boundingSphereRadius: Float
    get() = boundingVolume.boundingSphereRadius

open class TransformSpatial(val transform: Transform, _boundingVolume: AABB) : SimpleSpatial(_boundingVolume) {
    inline val center: Vector3f
        get() = boundingVolume.center
}
class StaticTransformSpatial(transform: Transform, val model: StaticModel) : TransformSpatial(transform, AABB(model.boundingVolume.localAABB)) {
    override fun update(deltaSeconds: Float) = boundingVolume.recalculate(transform)
}
// TODO: Is this still needed?
class AnimatedTransformSpatial(transform: Transform, model: AnimatedModel) : TransformSpatial(transform, AABB(model.boundingVolume.localAABB)) {
    override fun update(deltaSeconds: Float) = boundingVolume.recalculate(transform)
}
